lib/charms/oathkeeper/v0/forward_auth.py

```diff
@@ -0,0 +1,598 @@
+            )
+
+        if not relation or not relation.app:
+            return
+
+        data = _dump_data(ingress_app_names.to_dict(), FORWARD_AUTH_REQUIRER_JSON_SCHEMA)
+        relation.data[self.model.app].update(data)
+
+    def get_provider_info(self, relation_id: Optional[int] = None) -> Optional[ForwardAuthConfig]:
+        """Get the provider information from the databag."""
+        if len(self.model.relations) == 0:
+            return None
+        try:
+            relation = self.model.get_relation(self._relation_name, relation_id=relation_id)
+        except TooManyRelatedAppsError:
+            raise TooManyRelatedAppsError(
+                "More than one relation is defined. Please provide a relation_id"
+            )
+        if not relation or not relation.app:
+            return None
+
+        data = relation.data[relation.app]
+        if not data:
+            logger.debug("No relation data available.")
+            return
+
+        data = _load_data(data, FORWARD_AUTH_PROVIDER_JSON_SCHEMA)
+        forward_auth_config = ForwardAuthConfig.from_dict(data)
+        logger.debug(f"ForwardAuthConfig: {forward_auth_config}")
+
+        return forward_auth_config
+
+    def get_remote_app_name(self, relation_id: Optional[int] = None) -> Optional[str]:
+        """Get the remote app name."""
+        if len(self.model.relations) == 0:
+            return None
+        try:
+            relation = self.model.get_relation(self._relation_name, relation_id=relation_id)
+        except TooManyRelatedAppsError:
+            raise TooManyRelatedAppsError(
+                "More than one relation is defined. Please provide a relation_id"
+            )
+        if not relation or not relation.app:
+            return None
+
+        return relation.app.name
+
+    def is_ready(self, relation_id: Optional[int] = None) -> Optional[bool]:
+        """Checks whether ForwardAuth is ready on this relation.
+
+        Returns True when Oathkeeper shared the config; False otherwise.
+        """
+        if len(self.model.relations) == 0:
+            return None
+        try:
+            relation = self.model.get_relation(self._relation_name, relation_id=relation_id)
+        except TooManyRelatedAppsError:
+            raise TooManyRelatedAppsError(
+                "More than one relation is defined. Please provide a relation_id"
+            )
+
+        if not relation or not relation.app:
+            return None
+
+        return (
+            "decisions_address" in relation.data[relation.app]
+            and "app_names" in relation.data[relation.app]
+        )
+
+    def is_protected_app(self, app: str) -> bool:
+        if self.is_ready():
+            forward_auth_config = self.get_provider_info()
+            if forward_auth_config and app in forward_auth_config.app_names:
+                return True
+            return False
+
+        return False
```

Comment:
Reminder: This (and maybe some other changes) will need to be pushed to oathkeeper before merging this PR.

Reply:
Ref PR: https://github.com/canonical/oathkeeper-operator/pull/38